mlserver/batching/adaptive.py

```diff
@@ -0,0 +1,94 @@
+import time
+import asyncio
+
+from asyncio import Future, Queue, wait_for
+from typing import AsyncIterator, Awaitable, Dict, List
+
+from ..model import MLModel
+from ..types import (
+    InferenceRequest,
+    InferenceResponse,
+)
+
+from .requests import BatchedRequests
+
+
+class AdaptiveBatcher:
+    def __init__(self, model: MLModel):
+        self._model = model
+
+        self._max_batch_size = model.settings.max_batch_size
+        self._max_batch_time = model.settings.max_batch_time
+
+        # Save predict function before it gets decorated
+        self._predict_fn = model.predict
+        self._requests: Queue[InferenceRequest] = Queue(maxsize=self._max_batch_size)
+        self._async_responses: Dict[str, Future[InferenceResponse]] = {}
+        self._batching_task = None
+
+    async def predict(self, req: InferenceRequest) -> InferenceResponse:
+        await self._queue_request(req)
+        self._start_batcher_if_needed()
+        return await self._wait_response(req)
+
+    async def _queue_request(
+        self,
+        req: InferenceRequest,
+    ) -> Awaitable[InferenceResponse]:
+        await self._requests.put(req)
+
+        loop = asyncio.get_running_loop()
+        async_response = loop.create_future()
+
+        # TODO: What happens if ID is None?
```

Comment:
ok it could be none if we construct `InferenceRequest` manual and not passing an id as when we do the batching.

Reply:
Yes, exactly. The handlers should set if it's empty, although in SC it would be the executor the one ensuring that there's one set.

However, as @axsaucedo raised, it could be the case that the IDs are not unique. We were chatting of whether it would make sense to generate our own UIDs in the batching code, rather than relying on the request ones.

Reply:
Right, it should now be changed to use an internal set of IDs (i.e. without relying on the external request IDs).